src/ProtocolV3TestBase.sol

```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
     return IERC20(config.aToken).totalSupply() + supplyAmount > (config.supplyCap * 10 ** config.decimals);
   }
 
+  function _e2eTestLiquidation(
+    IPool pool,
+    address user,
+    address liquidator,
+    ReserveConfig memory collateralConfig,
+    ReserveConfig memory testAssetConfig,
+    uint256 amount
+  ) internal {
+    uint256 collateralAmountBefore = IERC20(collateralConfig.aToken).balanceOf(user);
+    this._borrow(testAssetConfig, pool, user, amount, false);
+    assertGt(IERC20(testAssetConfig.variableDebtToken).balanceOf(user), 0, 'DEBT_ZERO');
+    IPoolConfigurator configurator = IPoolConfigurator(
+      IPoolAddressesProvider(pool.ADDRESSES_PROVIDER()).getPoolConfigurator()
+    );
+    // set ltv/lt to 1bps which enables liquidation on the position
+    vm.prank(IPoolAddressesProvider(pool.ADDRESSES_PROVIDER()).getACLAdmin());
+    configurator.configureReserveAsCollateral(
+      collateralConfig.underlying,
+      1,
+      1,
+      10001
+    );
+    _liquidate(collateralConfig, testAssetConfig, pool, liquidator, user, amount, true);
+    // hf should be low enough to have the entire position be liquidated
+    assertEq(IERC20(testAssetConfig.variableDebtToken).balanceOf(user), 0, 'LIQUIDATION_DEBT_NOT_ZERO');
+    uint256 collateralAmountAfter = IERC20(collateralConfig.aToken).balanceOf(user);
+    assertLt(collateralAmountAfter, collateralAmountBefore, 'COLLATERAL_NO_DECREASE');
+    // within 1% error is fine
+    assertApproxEqRel(
+      IERC20(collateralConfig.aToken).balanceOf(liquidator),
+      collateralAmountBefore - collateralAmountAfter,
+      0.01e18,
+      'LIQUIDATOR_NO_RECEIVE_COLLATERAL'
+    );
+  }
+
+  function _e2eTestFlashLoan(
+    IPool pool,
+    ReserveConfig memory testAssetConfig,
+    uint256 amount
+  ) internal {
+    assertEq(IERC20(testAssetConfig.underlying).balanceOf(address(this)), 0, 'UNDERLYING_NOT_ZERO');
+    pool.flashLoanSimple(
+      address(this),
+      testAssetConfig.underlying,
+      amount,
+      abi.encode(address(pool)),
+      0
+    );
+    assertEq(IERC20(testAssetConfig.underlying).balanceOf(address(this)), 0, 'UNDERLYING_NOT_ZERO');
```

Comment:
I guess we can't really assert anything else here, all good - actually is there a fee for flashloans?

Reply:
Yes there is an optional fee. Aave has it turned on, Spark doesn't.